File: src/main/java/org/sandboxpowered/silica/client/resources/ResourceLoader.java
package org.sandboxpowered.silica.client.resources;

import org.sandboxpowered.silica.client.util.Identifier;

import java.io.IOException;
import java.io.InputStream;
import java.util.Set;
import java.util.function.Predicate;

public interface ResourceLoader {
    private static String getFilename(String type, Identifier identifier) {
        return String.format("%s/%s/%s", "assets", identifier.getNamespace(), identifier.getPath());
    }

    boolean containsFile(String path);

    InputStream openFile(String path) throws IOException;

    Set<String> findResources(String namespace, String path, int depth, Predicate<String> filter);

    Set<String> getNamespaces();
}

File: server/src/main/kotlin/org/sandboxpowered/silica/resources/ResourceLoader.kt
package org.sandboxpowered.silica.resources

import org.apache.commons.lang3.SystemUtils
import org.sandboxpowered.silica.api.util.Identifier
import java.io.File
import java.io.InputStream
import java.util.function.Predicate

interface ResourceLoader {
    val name: String

    fun contains(type: ResourceType, file: Identifier): Boolean

    fun open(file: String): InputStream

    fun open(type: ResourceType, file: Identifier): InputStream

    fun findResources(
        type: ResourceType,
        namespace: String? = null,
        category: String? = null,
        depth: Int = DEPTH_NO_LIMIT,
        filter: Predicate<String> = Predicate { true }
    ): Set<Identifier>

    fun getNamespaces(type: ResourceType): Set<String>

    companion object {
        fun getPath(type: ResourceType, identifier: Identifier): String {
            return "${type.folder}/${identifier.namespace}/${identifier.path}"
        }

        fun isValidPath(file: File, filename: String): Boolean {
            var string = file.canonicalPath
            if (SystemUtils.IS_OS_WINDOWS) {
                string = string.replace('\\', '/')
            }
            return string.endsWith(filename)
        }

        const val DEPTH_NO_LIMIT = -1
    }
}